keps/sig-apps/2255-pod-cost/README.md

```diff
@@ -55,102 +48,130 @@ Items marked with (R) are required *prior to targeting to a milestone / release*
+- Allow applications to influence the order of deleting pods when scaling down a ReplicaSet
 
 
 ### Non-Goals
 
-Guaranteed (in contrast to the recommendation stated in Goals) deletion of a selected replica.
+- Guarantees on pod deletion order
 
 ## Proposal
 
-The application can set the `controller.kubernetes.io/pod-cost` annotation to a pod through the Kubernetes API. When a downscale event happens, the pod with the lower priority value of the previously set annotation will be deleted first. If one pod of the Deployment has no priority annotation set, it will be treated as the lowest priority.
+Define a known annotation, namely `controller.kubernetes.io/pod-cost` that 
+applications can set to offer a hint on the cost of deleting a pod compared
+to other pods belonging to the same ReplicaSet. 
 
-If all pods have the same priority, there is no difference in the normal pod delete decision behaviour. The same applies if the pod-cost annotation is not used at all.
+The pod-cost range will be from [-MaxInt, MaxInt]. The default value is 0.
+Invalid values (like setting the annotation to string) will be converted to
+the default value.
 
-The pod-cost annotation can be changed during operation, for example, if workload changes or a new master gets elected.
+When scaling down a ReplicaSet, controller-manager will prioritize deleting
+pods with lower pod-cost. Specifically, the rank passed to 
+[ActivePodsWithRanks](https://github.com/kubernetes/kubernetes/blob/cac933934b1301665e6e51a81c66c483f4e16c49/pkg/controller/controller_utils.go#L784-L809)
+is pod-cost inverse then number of pods on node (the latter is the current 
+factor impacting the rank).
 
-### User Stories (optional)
+If none of the pods set the pod-cost annotation or all of them have the same value, then the 
+scale down behavior is not changed compared to now.
 
+### User Stories (optional)
 
 #### Story 1
 
-In an application environment with stateful worker (user-)sessions, it is essential to keep the user sessions alive as good as possible. In case of a scale-down event, the application has to tell the scheduler, which delete decision would have the lowest impact on existing sessions.
+The different pods of an application could have different utilization levels. 
+On scale down, the application may prefer to remove the pods with lower utilization.
 
 #### Story 2
 
-An application consists of identical server processes, but one of the replicas will be the master, which should be kept as long as possible. All other replicas can be treated as cattle workload. Then the master can set the priority annotation with a high priority value as soon as it has finished its startup process. The other replicas can remain either without any priority set, or e.g. with all the same, lower priority. This ensures, that the master replica of this deployment will be protected in a downscale situation.
+On scale down, the application may want to remove pods running on the most expensive 
```

Comment:
This is a completely different story from the previously written one, which is also valid. Why not add a new one?

Reply:
Is there a concrete example where this is valid? usually you don't even want to delete the master at all because that typically breaks the application, and so you want guarantees of deletion order which goes against the best effort nature of this feature.

Reply:
You are right, the case described should have a separate deployment for the master, with higher pod priority.

Reply:
There are some legacy applications, where this behavior is needed (master gets elected and should be kept)

Reply:
can you give an example?

Reply:
Some (legacy) databases

Reply:
ok, so a new replica is elected then if the master happens to be taken down, right?

Reply:
I am not too convinced that in those situations the deployment is actively scaled up and down. You usually define a replica count for databases, and this stays constant unless one goes down involuntarily, like a zone going offline.

Reply:
Aren't StatefulSets more appropriate for database use cases anyways?